Tests/NIOTests/SyscallAbstractionLayer.swift

```diff
@@ -0,0 +1,665 @@
+                                    file: StaticString = #file,
+                                    line: UInt = #line) throws -> SocketChannel {
+        let channel = try self.makeSocketChannel(eventLoop: self.loop)
+        let connectFuture = try channel.eventLoop.runSAL(syscallAssertions: {
+            try self.assertConnect(result: true, { $0 == remoteAddress })
+            try self.assertLocalAddress(address: localAddress)
+            try self.assertRemoteAddress(address: remoteAddress)
+            try self.assertRegister { selectable, eventSet, registration in
+                if case .socketChannel(let channel, let registrationEventSet) = registration {
+                    XCTAssertEqual(localAddress, channel.localAddress)
+                    XCTAssertEqual(remoteAddress, channel.remoteAddress)
+                    XCTAssertEqual(eventSet, registrationEventSet)
+                    XCTAssertEqual(.reset, eventSet)
+                    return true
+                } else {
+                    return false
+                }
+            }
+            try self.assertReregister { selectable, eventSet in
+                XCTAssertEqual([.reset, .readEOF], eventSet)
+                return true
+            }
+            // because autoRead is on by default
+            try self.assertReregister { selectable, eventSet in
+                XCTAssertEqual([.reset, .readEOF, .read], eventSet)
+                return true
+            }
+        }) {
+            channel.register().flatMap {
+                channel.connect(to: remoteAddress)
+            }
+        }
+        XCTAssertNoThrow(try connectFuture.teleportWait())
+        return channel
+    }
+
+    func tearDownSAL() {
+        SAL.printIfDebug("=== TEAR DOWN ===")
+        XCTAssertNotNil(self.kernelToUserBox)
+        XCTAssertNotNil(self.userToKernelBox)
+        XCTAssertNotNil(self.wakeups)
+        XCTAssertNotNil(self.group)
+
+        let group = DispatchGroup()
+        group.enter()
+        XCTAssertNoThrow(self.group.shutdownGracefully(queue: DispatchQueue.global()) { error in
+            XCTAssertNil(error, "unexpected error: \(error!)")
+            group.leave()
+        })
+        XCTAssertNoThrow(try self.assertWakeup())
+        group.wait()
+
+        self.group = nil
+        self.kernelToUserBox = nil
+        self.userToKernelBox = nil
+        self.wakeups = nil
+    }
+
+    func assertParkedRightNow(file: StaticString = #file, line: UInt = #line) throws {
+        SAL.printIfDebug("\(#function)")
+        let syscall = try self.userToKernelBox.waitForValue()
+        if case .whenReady = syscall {
+            return
+        } else {
+            XCTFail("unexpected syscall \(syscall)", file: file, line: line)
+        }
+    }
+
+    func assertWaitingForNotification(result: SelectorEvent<NIORegistration>?,
+                                      file: StaticString = #file, line: UInt = #line) throws {
+        SAL.printIfDebug("\(#function)")
```

Comment:
Any value in including the `result` here?

Reply:
Oops, looks like this broke your PR :)

Reply:
fixed